File: 143_very_hard_Sort_a_String_with_the_Given_Template.js
function customSort(t, s){

    //文字列tに基づいて優先順位をマッピングするオブジェクト
    const priority = {}
    t.split('').forEach((char, index) => {
        priority[char] = index;
    });

    // 文字列sをフィルタリングしてorderとoriginを作成
    const order = s.split('').filter(char => t.includes(char));
    const origin = s.split('').filter(char => !t.includes(char));

    // orderをカスタムソート
    const sortedOrder = order.sort((a, b) => {
        return (priority[a] != undefined ? priority[a] : -1) - (priority[b] !== undefined ? priority[b] : -1);
    });

    // originを通常の方法でソート
    const sortedOrigin = origin.sort();

    // 結果を結合
    return sortedOrder.join('') + sortedOrigin.join('');

}

console.log(customSort("edc", "abcdefzyx"));
console.log(customSort("fby", "abcdefzyx"));
console.log(customSort("", "abcdefzyx"));
console.log(customSort('cteqh', 'xnjanztmhg'));
console.log(customSort('jv', 'cxkafinfiuhnnaracsztbrcwaifwattzavwohoqskauififucq'));

File: 143_very_hard_Sort_a_String_with_the_Given_Template.py
"""
Sort a String with the Given Template
The function is given two strings t - template, s - to be sorted. Sort the characters in s such that if the character is present in t then it is sorted according to the order in t and other characters are sorted alphabetically after the ones found in t.

Examples
custom_sort("edc", "abcdefzyx") ➞ "edcabfxyz"

custom_sort("fby", "abcdefzyx") ➞ "fbyacdexz"

custom_sort("", "abcdefzyx") ➞ "abcdefxyz"

custom_sort("", "") ➞ ""
Notes
The characters in t and s are all lower-case.
"""

def custom_sort(t, s):

    order = "".join([i for i in s if i in t])
    origin = "".join([i for i in s if i not in t])

    priority = {char: i for i, char in enumerate(t)}

    def get_sort_key(x):
        return priority.get(x, -1)

    sorted_s = ''.join(sorted(order, key=get_sort_key))
    return sorted_s + ''.join(sorted(origin))

print(custom_sort("edc", "abcdefzyx"))
print(custom_sort("fby", "abcdefzyx"))
print(custom_sort("", "abcdefzyx"))
print(custom_sort('cteqh', 'xnjanztmhg'))
print(custom_sort('jv', 'cxkafinfiuhnnaracsztbrcwaifwattzavwohoqskauififucq'))